### DISCUSSION TITLE
[jailcheck] Warning: I can run programs in ...

### DISCUSSION THREAD

**Discussion by Rosika2** [netblue30/firejail#4398] [Q&A] [answered]
**Environment**
 - Linux distribution and version: Lubuntu 20.04.2 LTS
 - Firejail version: 0.9.66
 
Hi altogether,

running Linux Lubuntu 20.04.2 LTS I received my latest batch of updates yesterday, among them
firejail. So I´m on version 0.9.66 now.

I´ve learnt that there´s a new tool - **jailcheck** - available, which is phantastic. Thanks a lot for that.

However some programmes I run within firejail needed some tinkering with as they wouldn´t run the way they used to.

Finally I got **w3m**, **newsboat** and **podboat** working again.

For e.g. **podboat** I used the _default.profile_ and had just to comment out "include disable-programs.inc": **"#include disable-programs.inc"**. That way it works again.

But: when invoking the **jailcheck** command I get this:

```
4513:rosika::firejail podboat 
   Warning: AppArmor not enabled
   Virtual dirs: /var/tmp, 
   Warning: I can run programs in /home/rosika
   Warning: I can run programs in /tmp
   Warning: I can run programs in /run/user/1000
   Networking: enabled
```
   
... which got me wondering. How crucial are those **warnings**? Should I have done something different?

Comparing it to the results I got from "firejail firefox":

```
3102:rosika::firejail --private=/media/rosika/f14a27c2-0b49-4607-94ea-2e56bbf76fe1/DATEN-PARTITION/Dokumente/work2 firefox 
   Virtual dirs: /home/rosika, /tmp, /var/tmp, /dev, /usr/share, 
                 /run/user/1000, 
   Networking: enabled
```

I see I get no warnings here.

Could you help me somehow?

Thanks so much in advance.

Many greetings.
Rosika

**Comment by rusty-snake** [comment, marked as answer]
> Warning: I can run programs in …

That's because default.profile does not `include disbale-exec.in`. You can either create a podboat.profile and add it there or/and add `include disbale-exec.in` to default.local ([I do this](https://gitlab.com/rusty-snake/kyst/-/blob/f4d93bd61bdf9c044e36e73c398bb841864d6641/firejail/default.local#L7)).

> How crucial are those warnings?

Low. 

Having a `whitelist` profile and `dbus-{user,system} (filter|none)` is much more important.

Imagine an attacker can create a file in `/home/rosika` or `/tmp` containing malware and he/she can execute this file. Now, if you make this place `noexec` (via `disable-exec.inc`), he/she can no longer execute this file. However, maybe there are other directories which are writeable and don't have a `noexec` or the malware is written in python/perl/... or he/she has unrestricted access to the session-bus or etc. At this point you're quite lost. The only thing firejail provides here is to break the common ways how you do things so an attacker how want's to distribute his/her ransomware only cares about the 9 systems w/o firejail and ignores the one with firejail where things are complicated.

> For e.g. podboat I used the default.profile and had just to comment out "include disable-programs.inc": "#include disable-programs.inc". That way it works again.

Not recommended, default.profile is already weak and commenting `include disable-programs.inc`. Better is to copy default.profile to podboat.profile and add a `noblacklist` for that paths it need access to.
Usually you find the necessary `noblacklist`s with `firejail --build podboat` or `firejail --noprofile --name=podboat --private podboat`+`firejail --join=podboat tree -a`.

Or best: Copy [`/usr/share/doc/firejail/profile.template`](https://github.com/netblue30/firejail/blob/master/etc/templates/profile.template) to podboat.profile and start writing a profile for it (and open a PR with it :wink:).

> when invoking the jailcheck command I get this

Just to say it, applies for jailcheck as well:

https://github.com/rusty-snake/raudit/blob/4e913ebf291bcf749d92fe6f4771ef3764d74438/docs/source/how-can-i-fix.rst?plain=1#L8-L10

**Reply by Rosika2** [reply]
@rusty-snake:

Hi and thanks so much for your detailed reply.

> Better is to copy default.profile to podboat.profile and add a noblacklist for that paths it need access to.

O.K. That´s what I tried.
Yet it turns out getting the relevant `noblacklist`s is a bit hard to achieve.
`firejail --build podboat` didn´t work:
```

 firejail --build podboat
Starte podboat 2.18.0...
Lade Konfiguration...Fertig.
Error opening terminal: unknown.
Error: cannot run the sandbox
```
So I tired the second approach:
`firejail --noprofile --name=podboat --private podboat`
started an empty **podboat** .
Then: 
`firejail --join=podboat tree -a`
gave me this output:

```
firejail --join=podboat tree -a
Switching to pid 22352, the first child process inside the sandbox
Child process initialized in 14.31 ms
.
├── .bashrc
├── .config
│   └── fish
│       └── fish_variables
├── .local
│   └── share
│       └── fish
├── .newsboat
│   ├── pb-lock.pid
│   └── queue
└── .Xauthority

6 directories, 5 files
```
Yet I´m unsure what to do next. How can I get the repective paths I need to `noblacklist` from this output?

Tnx in advance.
Many greetings.
Rosika

**Reply by rusty-snake** [reply]
> Yet I´m unsure what to do next. How can I get the repective paths I need to noblacklist from this output?

1. Search for `blacklist`s for the paths.
```console
$ grep "config/fish" /etc/firejail/disable-*.inc 
/etc/firejail/disable-common.inc:read-only ${HOME}/.config/fish
$ grep "local/share/fish" /etc/firejail/disable-*.inc
/etc/firejail/disable-common.inc:deny-nolog  ${HOME}/.local/share/fish/fish_history
/etc/firejail/disable-common.inc:read-only ${HOME}/.local/share/fish
$ grep "newsboat" /etc/firejail/disable-*.inc
/etc/firejail/disable-programs.inc:deny  ${HOME}/.config/newsboat
/etc/firejail/disable-programs.inc:deny  ${HOME}/.local/share/newsboat
/etc/firejail/disable-programs.inc:deny  ${HOME}/.newsboat
```
2. Try if `noblacklist`ing the `blacklist`ed paths you found is enough.
```
noblacklist ${HOME}/.newsboat
```

**Reply by Rosika2** [reply]
Hi again @rusty-snake and thanks so much.

O.K., I added 

```
noblacklist ${HOME}/.config/fish
noblacklist ${HOME}/.local/share/fish/fish_history
noblacklist ${HOME}/.local/share/fish
noblacklist ${HOME}/.config/newsboat
noblacklist ${HOME}/.local/share/newsboat
noblacklist ${HOME}/.newsboat
```
to default.profile (of course I renamed it to `podboat.profile` first)
But it still doesn´t seem to work:

```
firejail podboat
Reading profile /home/rosika/.config/firejail/podboat.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-passwdmgr.inc
Reading profile /etc/firejail/disable-programs.inc
Warning: networking feature is disabled in Firejail configuration file
Parent pid 26847, child pid 26848
Warning: cleaning all supplementary groups
Child process initialized in 71.51 ms
XDG: Konfigurationsverzeichnis '/home/rosika/.config/newsboat' nicht zugänglich, benutze stattdessen '/home/rosika'.
Starte podboat 2.18.0...
Fehler: Eine Instanz von podboat läuft bereits (PID: 0)

Parent is shutting down, bye...
```
Curious: 
`Fehler: Eine Instanz von podboat läuft bereits (PID: 0)`
i.e. "an instance of "podboat" is already running", which simply isn´t true. I checked.

Many greetings.
Rosika

P.S.:

> Try if noblacklisting the blacklisted paths you found is enough.

Alas it´s not the case here.

**Reply by rusty-snake** [reply]
> Fehler: Eine Instanz von podboat läuft bereits (PID: 0)
i.e. "an instance of "podboat" is already running", which simply isn´t true. I checked.

If `~/.newsboat/pd-lock.pid` contains something like `pid:0`, try to `rm` it.

> i.e. "an instance of "podboat" is already running"

FYI: Start a program with e.g. `LANG=en_US.UTF8 LC_ALL=en_UTF.8` to get english errors.

> noblacklist ${HOME}/.config/fish
noblacklist ${HOME}/.local/share/fish/fish_history
noblacklist ${HOME}/.local/share/fish

Superfluous, as there are no `blacklist`s of them.

**Reply by Rosika2** [reply]
Hi again,

> If ~/.newsboat/pd-lock.pid contains something like pid:0, try to rm it.

The thing is: I don´t have that so I can´t remove anything.

```
ls -a ~/.newsboat
./  ../  cache.db
```
But `ps -aux | grep podboat` confirms that there´s **no** podboat-process running.

As I said in my first post:
`#include disable-programs.inc` within default.profile would get rid of any problems and let me perfectly run podboat.
But you told me it´s not recommended. 
Alas, until now everything I´ve tried resulted in a failure. Such a shame.

Tnx so much and greetings from Rosika.

P.S.:

> FYI: Start a program with e.g. LANG=en_US.UTF8 LC_ALL=en_UTF.8 to get english errors.

Thanks for the info. It´s highly appreciated.

**Reply by rusty-snake** [reply]
Maybe it has the lock-file somewhere else (`~/.cache/newboat`, `~/.config/newsboat`, `~/.local/share/newsboat`, `/run/user/$UID/.newsboat`, `/tmp/.newboat`). If not you can run with `--tracelog` and watch `journalctl --boot --pager-end --follow`. And if that's still don't help, I'm out of ideas for now.

**Reply by Rosika2** [reply]
Hi again,

O.K., I looked in all the possible paths you provided. The lock-file definitively **doesn´t** exist anywhere.

As far as your latest recommendation is concerned (--tracelog) I did what you suggested.
Thanks a lot for that great idea.

Here´s what **journalctl** came up with:

```
Jul 13 18:16:08 rosika-10159 firejail[33931]: blacklist violation - sandbox 33920, exe podboat, syscall access, path /home/rosika/.config/newsboat
Jul 13 18:16:08 rosika-10159 firejail[33931]: blacklist violation - sandbox 33920, exe podboat, syscall mkdir, path /home/rosika/.newsboat
```
Hmm, I´m not really sure how to interpret that correctly.

But thanks so much for your help.
Many greetings.
Rosika

**Reply by rusty-snake** [reply]
Sounds like they are still `blacklist`ed. 
- Did you add the `noblacklist`s before the `include disable-programs.inc`? (this is the right order)
- What does `firejail --profile=podboat ls -l ~/.newboat` show?

**Reply by Rosika2** [reply]
Hi once more.

Well, I´ll be blessed.

> Did you add the noblacklists before the include disable-programs.inc? (this is the right order)

You were so very right in assuming I did it the wrong way.
In fact I had the `noblacklist`s after the `include disable-programs.inc`. I placed them before the entire **include** block and now it works perfectly.

In fact I didn´t know the order was of vital importance. So sorry to have been such a driving force in wasting your time. 
I sincerely apologize.

Nevertheless I´m so grateful for all of your help which is so much appreciated.
Many, many thanks again. 

So many greetings from Rosika.

P.S.:

`firejail --profile=podboat ls -l ~/.newsboat` yielded th following (after setting things straight):
 
```
firejail --profile=podboat ls -l ~/.newsboat
Reading profile /home/rosika/.config/firejail/podboat.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-passwdmgr.inc
Reading profile /etc/firejail/disable-programs.inc
Warning: networking feature is disabled in Firejail configuration file
Parent pid 36924, child pid 36926
Warning: cleaning all supplementary groups
Child process initialized in 71.88 ms
insgesamt 56
-rw-r--r-- 1 rosika rosika 53248 Feb 25 14:54 cache.db

Parent is shutting down, bye...

```

**Reply by rusty-snake** [reply]
> In fact I didn´t know the order was of vital importance.

I think order is probably the second most difficult point for novice users. Not only that it matter in which order you add commands, some commands have a first-win logic and others a last-win logic.

> now it works perfectly

:confetti_ball:

**Reply by Rosika2** [reply]
@rusty-snake:

Hi again,

thanks for the additional comments.
In the meantime I took a look at the `profile.template`:

```
Sections structure
#   HEADER
#   COMMENTS
#   IGNORES
#   NOBLACKLISTS
#   ALLOW INCLUDES
#   BLACKLISTS
#   DISABLE INCLUDES
[...]

```
So I could (or rather should) have learnt by that about the correct order of entries :) 

Running **podboat** with the new and correct configuration I decided to take a look at 
`sudo jailcheck` once more:

```

4728:rosika::firejail podboat 
   Warning: AppArmor not enabled
   Virtual dirs: /var/tmp, 
   Warning: I can run programs in /home/rosika
   Warning: I can run programs in /tmp
   Warning: I can run programs in /run/user/1000
   Networking: enabled

```
So I still get the "warnings".
Yet in your first reply you said:

>     How crucial are those warnings?
> 
> Low.
> Having a whitelist profile and dbus-{user,system} (filter|none) is much more important.

So I took a look at my podboat.profile.  I´ve taken care of the `noblacklist`s but the other point you referred to (dbus) looks like this (as I´ve taken the default.profile as a starting point):

```
# dbus-user none
# dbus-system none

```

Should there be done anything about this?

Sorry for my new question.

Thanks so much.
Many greetings.
Rosika

**Reply by rusty-snake** [reply]
> So I still get the "warnings".

> That's because default.profile does not `include disbale-exec.in`. You can either create a podboat.profile and add it there or/and add `include disbale-exec.in` to default.local ([I do this](https://gitlab.com/rusty-snake/kyst/-/blob/f4d93bd61bdf9c044e36e73c398bb841864d6641/firejail/default.local#L7)).

> Should there be done anything about this?

General: 
- You can uncomment the most commands in podboat.profile without breaking anything.
- You should uncomment the most commands in podboat.profile.

Specific:
- Yes I don't expect that they break podboat and D-Bus is probably the biggest sandbox escape vector.

**Reply by Rosika2** [reply]
@rusty-snake:

Hi again and thank you very much for your help.

> That's because default.profile does not include disbale-exec.in. You can either create a podboat.profile and add it there or/and add include disbale-exec.in to default.local (I do this).

Of course, you´re right. You already told me so  in your first response. Thanks again for pointing that out.

> Yes I don't expect that they break podboat and D-Bus is probably the biggest sandbox escape vector.

Great. I´ve done that/uncommented the following:


```
dbus-user none
dbus-system none
```
And podboat works great that way.
Well, I´ve learnt a lot with your help and I hope to be able to apply some of my newly acquired knowledge in the future.

Many thanks again. Keep safe and many greetings.
Rosika

**Reply by curiosityseeker** [reply]
> 
> I think order is probably the second most difficult point for novice users. Not only that it matter in which order you add commands, some commands have a first-win logic and others a last-win logic.

Is this somewhere properly documented other than in `profile.template`and [here](https://github.com/netblue30/firejail/wiki/Creating-Profiles#guidelines)? I'm asking this as I'm about to overhaul the Firejail article in the Arch Linux wiki.

**Reply by rusty-snake** [reply]
> Is this somewhere properly documented other than in profile.template and here?

No, profile.template is probably the best place after source-code reading/testing.

> I'm asking this as I'm about to overhaul the Firejail article in the Arch Linux wiki.

:rocket:

**Reply by curiosityseeker** [reply]
> No, profile.template is probably the best place after source-code reading/testing.

Thanks, that's what I thought.

I've modified many entries in the Firejail article in the Arch Linux wiki and added a lot of stuff. But I guess there is still a lot to do. Would you mind looking through the site if time permits? Hints regarding errors and missing things would be helpful.

> rocket

Branson - Bezos - curiosityseeker. Ha!

**Reply by rusty-snake** [reply]
My thoughts while reading the changes, some are good and useful and some can be just ignored :slightly_smiling_face:.

<details><summary>click me</summary>

#### How to remove specific Firejail symbolic links

> If you don't want to use Firejail for a specific application […] you have to manually remove the related symbolic link _after every call of `firecfg`_:

Relevant issues:
 - [firecfg: allow for ignoring specific apps](https://github.com/netblue30/firejail/issues/2097)
 - [With firecfg, how do I configure specific applications to go through firejail?](https://github.com/netblue30/firejail/issues/3665)

#### Use with hardened_malloc in a custom profile.

I'm not sure about

```
env LD_PRELOAD='/usr/lib/libhardened_malloc.so'
```

but I think

```
env LD_PRELOAD=/usr/lib/libhardened_malloc.so
```

is the right syntax.

#### Removing the "Desktop Files" section and merging with "Using Firejail by default"

1. `firecfg` does not create symlinks for all program with a profile. Only programs
   in firecfg.config or with a profile in `~/.config/firejail` will get firejailed.
2. `firecfg --fix` does not create symlinks, you still need to create symlinks.
   Maybe with `firecfg --bindir=~/.local/bin`.
3. `firecfg` has only a partial GNOME support, see [firecfg dont detect all .desktop files for cleaning](https://github.com/netblue30/firejail/issues/2624) for example.
4. Executing `firecfg --fix` as root creates the fixed .desktop files in /root, it don't?
5. It does two checks, one for Exec=/abs/path and one for DBusActivable=true
6. and it does firejail.users and apparmor stuff

#### pacman hook: change "firecfg" to "firecfg --fix"

This looks wrong to me. Given that

> Description = Configure symlinks in /usr/local/bin based on firecfg.config...

#### Rephrasing explanations for blacklists and whitelists, adding more explanations and examples

> Deny access to a directory or file and permit everything else

It does not really permit, it just don't say anything about every other path.

> Disable or undo

Maybe it makes sense to move them out so they are not in the explanation of an other command.
Instead only one "to disable/undo/ignore a black/whitelist prefix it with a no" paragraph.

> must be added above

It's a virtual above because of `include`.
I usually say 'before', however I'm not a native english speaker.

> and must add a whitelist directive

_OT: We need a consistent naming, command/option/directive/flag/..._

Maybe a "Note" block with whitelist restrictions makes sense. e.g.
- You can not rename a whitelisted paths, however you can rename files in a whitlisted path. (KDE apps often have immutable settings because of that)
- You can only whitelist paths that exist, use mkdir/mkfile.

#### Added a "Hardening Firejail" section

In general, nnp breaks any privileged program (all without nnp in it's profile are possible affected).
Notable are vbox, wireshark, chrom*  with linux-hardened

`restricted-network yes` and firejail.users can be "suggested" here to.
firejail-welcome.sh can set both, force-nnp and restricted-network.

</details>

**Reply by curiosityseeker** [reply]
> #### How to remove specific Firejail symbolic links
> 
> > If you don't want to use Firejail for a specific application […] you have to manually remove the related symbolic link _after every call of `firecfg`_:
> 
> Relevant issues:
> 
>     * [firecfg: allow for ignoring specific apps](https://github.com/netblue30/firejail/issues/2097)
> 
>     * [With firecfg, how do I configure specific applications to go through firejail?](https://github.com/netblue30/firejail/issues/3665)

Yes, I have to think about how to best present it. The problem with removing the symlink is also that it will be re-added when using the pacman hook. So commenting the respective application in `firecfg.config`should be added.

> 
> 
> #### Use with hardened_malloc in a custom profile.
> 
> I'm not sure about
> 
> ```
> env LD_PRELOAD='/usr/lib/libhardened_malloc.so'
> ```
> 
> but I think
> 
> ```
> env LD_PRELOAD=/usr/lib/libhardened_malloc.so
> ``
> 
> is the right syntax.

I had tried it once and it worked.

> #### Removing the "Desktop Files" section and merging with "Using Firejail by default"
> 
>     1. `firecfg` does not create symlinks for all program with a profile. Only programs
>        in firecfg.config or with a profile in `~/.config/firejail` will get firejailed.

Yes, it's mentioned in the "Note" block in that section but it's probably better to move it to the firecfg instructions.

> 
>     2. `firecfg --fix` does not create symlinks, you still need to create symlinks.
>        Maybe with `firecfg --bindir=~/.local/bin`.

Yes, you're right. That's definitely a problem. The correct way without sudo would be executing `firecfg` as root and `firecfg --fix` as user.
> 
>     3. `firecfg` has only a partial GNOME support, see [firecfg dont detect all .desktop files for cleaning](https://github.com/netblue30/firejail/issues/2624) for example.

Thanks, I wasn't aware of that.
> 
>     4. Executing `firecfg --fix` as root creates the fixed .desktop files in /root, it don't?

See above.
> 
>     5. It does two checks, one for Exec=/abs/path and one for DBusActivable=true
> 
>     6. and it does firejail.users and apparmor stuff
> 
> 
> #### pacman hook: change "firecfg" to "firecfg --fix"
> 
> This looks wrong to me. Given that
> 
Yes, right. 
> 
> #### Rephrasing explanations for blacklists and whitelists, adding more explanations and examples
> 
> > Deny access to a directory or file and permit everything else
> 
> It does not really permit, it just don't say anything about every other path.

Hm, yes, but don't you think that this phrasing is acceptable to illustrate the point? Any other suggestions?

> 
> > Disable or undo
> 
> Maybe it makes sense to move them out so they are not in the explanation of an other command.
> Instead only one "to disable/undo/ignore a black/whitelist prefix it with a no" paragraph.

I will consider this.

> 
> > must be added above
> 
> It's a virtual above because of `include`.
> I usually say 'before', however I'm not a native english speaker.

Neither am I ;-)

> 
> _OT: We need a consistent naming, command/option/directive/flag/..._

Yes, but I think even in the Firejail documentation there is no consistent wording, is it?

> 
> Maybe a "Note" block with whitelist restrictions makes sense. e.g.
> 
>     * You can not rename a whitelisted paths, however you can rename files in a whitlisted path. (KDE apps often have immutable settings because of that)
> 
>     * You can only whitelist paths that exist, use mkdir/mkfile.

I'm not quite sure that I understand what you're after ...
> 
> 
> #### Added a "Hardening Firejail" section
> 
> In general, nnp breaks any privileged program (all without nnp in it's profile are possible affected).
> Notable are vbox, wireshark, chrom* with linux-hardened

Yes, probably because that kernel doesn't support unprivileged user namespaces. With linux, linux-zen and linux-lts only Vitualbox seems to be affected.
> 
> `restricted-network yes` and firejail.users can be "suggested" here to.
> firejail-welcome.sh can set both, force-nnp and restricted-network.

Thanks, I'll consider that.

Thanks a lot for your comprehensive answer! Much appreciated!

**Reply by rusty-snake** [reply]
> > It does not really permit, it just don't say anything about every other path.
>
> Hm, yes, but don't you think that this phrasing is acceptable to illustrate the point? Any other suggestions?

As I said, just thoughts

> > _OT: We need a consistent naming, command/option/directive/flag/..._
>
> Yes, but I think even in the Firejail documentation there is no consistent wording, is it?

It's missing everywhere.

> I'm not quite sure that I understand what you're after ...

```wiki
{{Note|<nowiki></nowiki>
* You can not rename a whitelisted paths, however you can rename files in a whitlisted path. (KDE apps often have immutable settings because of that). See issue #123456
* You can only whitelist paths that exist, use mkdir/mkfile.
}}
```

**Reply by curiosityseeker** [reply]
> ```
> {{Note|<nowiki></nowiki>
> * You can not rename a whitelisted paths, however you can rename files in a whitlisted path. (KDE apps often have immutable settings because of that). See issue #123456
> * You can only whitelist paths that exist, use mkdir/mkfile.
> }}
> ```
Ah, okay, I now see what you mean! But sorry - I still don't really get the first issue ... Could you, please, illustrate what it's about?

**Reply by rusty-snake** [reply]
Relevant issues: #1793 and #2874

If you have `whitelist ~/.mozilla`, you can not `mv ~/.mozilla foo` or `mv foo ~/.mozilla`. What you can it `mv ~/.mozilla/bar foo` or `mv foo ~/.mozilla/bar`. That's problematic if you `whitelist` a file and the program writes to the file by 1. create file.858674 and write to it 2. rename file.858674 to file.
